container/view.go

```diff
@@ -0,0 +1,288 @@
+			snapshot.ExposedPorts[exposed] = struct{}{}
+		}
+	}
+
+	if len(container.Args) > 0 {
+		args := []string{}
+		for _, arg := range container.Args {
+			if strings.Contains(arg, " ") {
+				args = append(args, fmt.Sprintf("'%s'", arg))
+			} else {
+				args = append(args, arg)
+			}
+		}
+		argsAsString := strings.Join(args, " ")
+		snapshot.Command = fmt.Sprintf("%s %s", container.Path, argsAsString)
+	} else {
+		snapshot.Command = container.Path
+	}
+
+	if container.NetworkSettings != nil {
+		networks := make(map[string]*network.EndpointSettings)
+		for name, netw := range container.NetworkSettings.Networks {
+			if netw == nil || netw.EndpointSettings == nil {
+				continue
+			}
+			networks[name] = &network.EndpointSettings{
+				EndpointID:          netw.EndpointID,
+				Gateway:             netw.Gateway,
+				IPAddress:           netw.IPAddress,
+				IPPrefixLen:         netw.IPPrefixLen,
+				IPv6Gateway:         netw.IPv6Gateway,
+				GlobalIPv6Address:   netw.GlobalIPv6Address,
+				GlobalIPv6PrefixLen: netw.GlobalIPv6PrefixLen,
+				MacAddress:          netw.MacAddress,
+				NetworkID:           netw.NetworkID,
+			}
+			if netw.IPAMConfig != nil {
+				networks[name].IPAMConfig = &network.EndpointIPAMConfig{
+					IPv4Address: netw.IPAMConfig.IPv4Address,
+					IPv6Address: netw.IPAMConfig.IPv6Address,
+				}
+			}
+		}
+		snapshot.NetworkSettings = &types.SummaryNetworkSettings{Networks: networks}
+		for port, bindings := range container.NetworkSettings.Ports {
+			p, err := nat.ParsePort(port.Port())
+			if err != nil {
+				logrus.Warnf("invalid port map %+v", err)
+				continue
+			}
+			if len(bindings) == 0 {
+				snapshot.Ports = append(snapshot.Ports, types.Port{
+					PrivatePort: uint16(p),
+					Type:        port.Proto(),
+				})
+				continue
+			}
+			for _, binding := range bindings {
+				h, err := nat.ParsePort(binding.HostPort)
+				if err != nil {
+					logrus.Warnf("invalid host port map %+v", err)
+					continue
+				}
+				snapshot.Ports = append(snapshot.Ports, types.Port{
+					PrivatePort: uint16(p),
+					PublicPort:  uint16(h),
+					Type:        port.Proto(),
+					IP:          binding.HostIP,
+				})
+			}
+		}
+	}
+	return snapshot
+}
+
+// EmbeddedStringFieldIndex is used to extract a field from an embedded struct
+// using reflection and builds an index on that field.
+type EmbeddedStringFieldIndex struct {
+	EmbeddedStruct string
+	Field          memdb.StringFieldIndex
+}
+
+// FromObject implements the memdb.SingleIndexer interface
+// object or to indicate that the index value is missing.
+func (e *EmbeddedStringFieldIndex) FromObject(obj interface{}) (bool, []byte, error) {
+	v := reflect.ValueOf(obj)
+	v = reflect.Indirect(v) // Dereference the pointer if any
+
+	ev := v.FieldByName(e.EmbeddedStruct)
+	if !ev.IsValid() {
+		return false, nil,
+			fmt.Errorf("embedded struct '%s' for %#v is invalid", e.EmbeddedStruct, obj)
+	}
+	return e.Field.FromObject(ev.Interface())
```

Comment:
Why is reflection necessary here? It looks like this is only used to index by ID. Couldn't this function be written to just return the ID from a `Snapshot` struct? (see for example https://github.com/docker/swarmkit/blob/bump_v1.13.2/manager/state/store/clusters.go#L192-L211)